File: Controles.cpp
#include "Motor.h"
#include "Arduino.h"
#include "Controles.h"
#include <Wire.h>
#include <Adafruit_GFX.h>
#include <Adafruit_SSD1306.h>

int dir;
int lastDir;

Controles::Controles(int inputSubir, int inputDescer, int inputAlturaCustom, Motor *esquerda, Motor *direita) {
  this->_inputSubir = inputSubir;
  this->_inputDescer = inputDescer;
  this->_motorEsquerda = esquerda;
  this->_motorDireita = direita;
  this->_lastStep = 0;
  this->_direcao = PARAR;
  this->_serialComando = PARAR;
  this->_calibrando = false;
  this->_ultimaSincronizacao = 0;
  this->_erroSincAtual = 0;
  this->_recuperando = false;
  this->_stepsEInicio = 0;
  this->_stepsDInicio = 0;
  this->_modoPreset = false;
  this->_alvoEmCm = 0.0f;
}

void Controles::begin() {
  pinMode(this->_inputSubir, INPUT);
  pinMode(this->_inputDescer, INPUT);
}

void Controles::loop() {
  if (_calibrando) return;

  bool isSubir = digitalRead(this->_inputSubir);
  bool isDescer = digitalRead(this->_inputDescer);

  // Botao fisico cancela recuperacao
  if (_recuperando && (isSubir || isDescer)) {
    _recuperando = false;
    _motorEsquerda->parar();
    _motorDireita->parar();
  }

  // Durante recuperacao, volta para posicao inicial
  if (_recuperando) {
    int stepsE = _motorEsquerda->getSteps();
    if (stepsE > _stepsEInicio) {
      _motorEsquerda->descer();
      _motorDireita->descer();
      dir = -1;
    } else if (stepsE < _stepsEInicio) {
      _motorEsquerda->subir();
      _motorDireita->subir();
      dir = 1;
    } else {
      _motorEsquerda->parar();
      _motorDireita->parar();
      _recuperando = false;
      dir = 0;
      Serial.println("{\"tipo\":\"info\",\"msg\":\"recuperacao_concluida\"}");
    }
    this->_verificaDirecao();
    return;
  }

  // Modo preset: mover para altura alvo
  if (_modoPreset) {
    if (isSubir || isDescer) {
      _modoPreset = false;
      _motorEsquerda->parar();
      _motorDireita->parar();
      dir = 0;
      this->_verificaDirecao();
      return;
    }
    float posAtual = _motorEsquerda->getPosicaoEmCentimetros();
    float diff = _alvoEmCm - posAtual;
    if (abs(diff) < PRESET_TOLERANCIA_CM) {
      _motorEsquerda->parar();
      _motorDireita->parar();
      _modoPreset = false;
      dir = 0;
      char json[80];
      snprintf(json, sizeof(json),
        "{\"tipo\":\"info\",\"msg\":\"preset_concluido\",\"posicao\":%.2f}", posAtual);
      Serial.println(json);
    } else if (diff > 0) {
      if (_direcao == PARAR) {
        _stepsEInicio = _motorEsquerda->getSteps();
        _stepsDInicio = _motorDireita->getSteps();
      }
      _motorEsquerda->subir();
      _motorDireita->subir();
      dir = 1;
    } else {
      if (_direcao == PARAR) {
        _stepsEInicio = _motorEsquerda->getSteps();
        _stepsDInicio = _motorDireita->getSteps();
      }
      _motorEsquerda->descer();
      _motorDireita->descer();
      dir = -1;
    }
    if (millis() - _ultimaSincronizacao >= 20) {
      _sincronizarMotores();
      _ultimaSincronizacao = millis();
    }
    this->_verificaDirecao();
    return;
  }

  if (isSubir && !isDescer) {
    if (_direcao == PARAR) {
      _stepsEInicio = _motorEsquerda->getSteps();
      _stepsDInicio = _motorDireita->getSteps();
    }
    _serialComando = PARAR;  // botao cancela comando serial
    this->_motorEsquerda->subir();
    this->_motorDireita->subir();
    dir = 1;
  } else if (!isSubir && isDescer) {
    if (_direcao == PARAR) {
      _stepsEInicio = _motorEsquerda->getSteps();
      _stepsDInicio = _motorDireita->getSteps();
    }
    _serialComando = PARAR;  // botao cancela comando serial
    this->_motorEsquerda->descer();
    this->_motorDireita->descer();
    dir = -1;
  } else if (_serialComando == SUBIR) {
    this->_motorEsquerda->subir();
    this->_motorDireita->subir();
    dir = 1;
  } else if (_serialComando == DESCER) {
    this->_motorEsquerda->descer();
    this->_motorDireita->descer();
    dir = -1;
  } else {
    this->_motorEsquerda->parar();
    this->_motorDireita->parar();
    dir = 0;
  }

  if (millis() - _ultimaSincronizacao >= 20) {
    _sincronizarMotores();
    _ultimaSincronizacao = millis();
  }

  this->_verificaDirecao();
}

void Controles::_verificaDirecao() {
  int currentStep = this->_motorEsquerda->getSteps();
  DIRECAO anterior = this->_direcao;

  if (currentStep > this->_lastStep) {
    this->_direcao = SUBIR;
  } else if (currentStep < this->_lastStep) {
    this->_direcao = DESCER;
  } else {
    this->_direcao = PARAR;
  }

  // Salva posicao na NVS ao parar
  if (anterior != PARAR && this->_direcao == PARAR) {
    _motorEsquerda->saveSteps();
    _motorDireita->saveSteps();
  }

  this->_lastStep = currentStep;
}

void Controles::subir() {
  _stepsEInicio = _motorEsquerda->getSteps();
  _stepsDInicio = _motorDireita->getSteps();
  _recuperando = false;
  _serialComando = SUBIR;
  this->_motorEsquerda->subir();
  this->_motorDireita->subir();
}

void Controles::descer() {
  _stepsEInicio = _motorEsquerda->getSteps();
  _stepsDInicio = _motorDireita->getSteps();
  _recuperando = false;
  _serialComando = DESCER;
  this->_motorEsquerda->descer();
  this->_motorDireita->descer();
}

void Controles::parar() {
  _serialComando = PARAR;
  _modoPreset = false;
  this->_motorEsquerda->parar();
  this->_motorDireita->parar();
}

void Controles::irPara(float alvoEmCm) {
  if (alvoEmCm < 84.0f || alvoEmCm > 144.0f) {
    Serial.println("{\"tipo\":\"erro\",\"msg\":\"preset_fora_de_faixa\"}");
    return;
  }
  _alvoEmCm = alvoEmCm;
  _stepsEInicio = _motorEsquerda->getSteps();
  _stepsDInicio = _motorDireita->getSteps();
  _modoPreset = true;
  _recuperando = false;
  char json[72];
  snprintf(json, sizeof(json),
    "{\"tipo\":\"info\",\"msg\":\"preset_iniciado\",\"alvo\":%.1f}", alvoEmCm);
  Serial.println(json);
}

void Controles::custom() {
}

DIRECAO Controles::getDirecao() const {
  return _direcao;
}

bool Controles::isBotaoSubir() const {
  return digitalRead(_inputSubir);
}

bool Controles::isBotaoDescer() const {
  return digitalRead(_inputDescer);
}

int Controles::getErroSincPulsos() const {
  return _erroSincAtual;
}

void Controles::_sincronizarMotores() {
  if (_recuperando) return;

  if (_direcao == PARAR) {
    _motorEsquerda->setVelocidade(255);
    _motorDireita->setVelocidade(255);
    _erroSincAtual = 0;
    return;
  }

  int stepsE = _motorEsquerda->getSteps();
  int stepsD = _motorDireita->getSteps();
  int pulsosE = _motorEsquerda->getPulsosCalibrados();
  int pulsosD = _motorDireita->getPulsosCalibrados();

  float posNormE = (float)stepsE / pulsosE;
  float posNormD = (float)stepsD / pulsosD;
  float erro = posNormE - posNormD;

  _erroSincAtual = (int)(erro * 6800);

  if (abs(_erroSincAtual) > EMERGENCIA_PULSOS) {
    _motorEsquerda->parar();
    _motorDireita->parar();
    _serialComando = PARAR;
    _recuperando = true;
    _modoPreset = false;
    // Distingue encoder inativo (um lado zerado) de desincronizacao real
    bool encoderInativo = (stepsD == _stepsDInicio) || (stepsE == _stepsEInicio);
    if (encoderInativo) {
      Serial.println("{\"tipo\":\"erro\",\"msg\":\"encoder_inativo\"}");
    } else {
      Serial.println("{\"tipo\":\"erro\",\"msg\":\"sinc_emergencia\"}");
    }
    return;
  }

  if (erro > 0.0005f) {
    int velE = constrain((int)(255.0f - GANHO_SINC * erro), VEL_MIN_SINC, 255);
    _motorEsquerda->setVelocidade(velE);
    _motorDireita->setVelocidade(255);
  } else if (erro < -0.0005f) {
    int velD = constrain((int)(255.0f + GANHO_SINC * erro), VEL_MIN_SINC, 255);
    _motorEsquerda->setVelocidade(255);
    _motorDireita->setVelocidade(velD);
  } else {
    _motorEsquerda->setVelocidade(255);
    _motorDireita->setVelocidade(255);
  }
}

void Controles::calibrar(Adafruit_SSD1306 *tela) {
  _calibrando = true;
  _serialComando = PARAR;

  // 1. Exibir mensagem e publicar status
  tela->clearDisplay();
  tela->setCursor(0, 0);
  tela->setTextSize(1);
  tela->println("Calibrando...");
  tela->println("Descendo...");
  tela->display();
  Serial.println("{\"tipo\":\"status\",\"status\":\"calibrando\"}");

  // 2. Descer ambos os motores
  _motorEsquerda->descer();
  _motorDireita->descer();

  // 3. Aguardar ate ambos travarem (fim de curso inferior)
  delay(2000);  // tempo minimo para comecar a mover
  while (!(_motorEsquerda->isTravado(800) && _motorDireita->isTravado(800))) {
    delay(50);
  }

  // 4. Parar e zerar steps
  _motorEsquerda->parar();
  _motorDireita->parar();
  delay(300);
  _motorEsquerda->setSteps(0);
  _motorDireita->setSteps(0);

  // 5. Exibir mensagem de subida
  tela->clearDisplay();
  tela->setCursor(0, 0);
  tela->println("Subindo para");
  tela->println("medir...");
  tela->display();
  Serial.println("{\"tipo\":\"status\",\"status\":\"calibrando_subindo\"}");

  delay(500);

  // 6. Subir ambos os motores
  _motorEsquerda->subir();
  _motorDireita->subir();

  // 7. Aguardar ate ambos travarem (fim de curso superior)
  delay(2000);
  while (!(_motorEsquerda->isTravado(800) && _motorDireita->isTravado(800))) {
    delay(50);
  }

  // 8. Registrar pulsos medidos
  int pulsosE = _motorEsquerda->getSteps();
  int pulsosD = _motorDireita->getSteps();

  // 9. Parar motores
  _motorEsquerda->parar();
  _motorDireita->parar();

  // 10. Salvar via Preferences (NVS)
  _motorEsquerda->setPulsosCalibrados(pulsosE);
  _motorDireita->setPulsosCalibrados(pulsosD);
  _motorEsquerda->saveSteps();
  _motorDireita->saveSteps();

  // 11. Exibir resultado e publicar
  tela->clearDisplay();
  tela->setCursor(0, 0);
  tela->println("Calibrado!");
  char buf[32];
  snprintf(buf, sizeof(buf), "E:%d D:%d", pulsosE, pulsosD);
  tela->println(buf);
  tela->display();

  char json[96];
  snprintf(json, sizeof(json),
    "{\"tipo\":\"calibracao\",\"pulsosE\":%d,\"pulsosD\":%d}",
    pulsosE, pulsosD);
  Serial.println(json);

  _calibrando = false;
}

void Controles::printStatus(Adafruit_SSD1306 *tela) {
  tela->clearDisplay();
  tela->setCursor(0, 0);
  tela->setTextSize(2);

  if (this->_direcao == SUBIR) {
    tela->println("Subindo");
  } else if (this->_direcao == DESCER) {
    tela->println("Descendo");
  } else {
    tela->println("Parado");
  }

  tela->setTextSize(4);

  char total[16];
  float posicaoEmCentimetros = this->_motorEsquerda->getPosicaoEmCentimetros();
  dtostrf(posicaoEmCentimetros, 3, 2, total);

  char *cm = strtok(total, ".");
  tela->print(cm);
  tela->println("cm");

  char *mm = strtok(NULL, ".");
  tela->setTextSize(1);
  tela->print(mm);
  tela->print("mm");

  tela->printf(" S:%d", _erroSincAtual);

  tela->setCursor(0, 56);
  tela->setTextSize(1);
  tela->print(digitalRead(_inputSubir) ? "[^]" : " ^ ");
  tela->print(digitalRead(_inputDescer) ? "[v]" : " v ");

  tela->display();
}

File: rpi3/src/Controles.ts
import { Motor, Direcao } from './Motor';
import { MqttClient, MqttComando } from './MqttClient';
import { SINC, PRESET, MOTOR } from './config';
import * as fs from 'fs/promises';
import { CALIBRACAO_PATH } from './config';

export interface StatusPayload {
  status: string;
  posicao: number;
  stepsE: number;
  stepsD: number;
  erroSync: number;
  velE: number;
  velD: number;
  btnS: boolean;
  btnD: boolean;
}

export class Controles {
  private readonly _motorEsquerda: Motor;
  private readonly _motorDireita: Motor;
  private readonly _pinSubir: number;
  private readonly _pinDescer: number;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  private readonly _pig: any;

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  private _gpioBtnSubir!: any;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  private _gpioBtnDescer!: any;

  // Estado cacheado dos botões (atualizado via notify, lido de forma síncrona no loop)
  private _btnSubirLevel: number = 0;
  private _btnDescerLevel: number = 0;

  private _mqttComando: MqttComando = 'PARAR';
  private _calibrando: boolean = false;
  private _recuperando: boolean = false;
  private _modoPreset: boolean = false;
  private _alvoEmCm: number = 0;

  private _stepsEInicio: number = 0;
  private _stepsDInicio: number = 0;
  private _erroSincAtual: number = 0;

  private _pollInterval: ReturnType<typeof setInterval> | null = null;
  private _sincInterval: ReturnType<typeof setInterval> | null = null;

  constructor(
    pinSubir: number,
    pinDescer: number,
    motorEsquerda: Motor,
    motorDireita: Motor,
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    pig: any,
  ) {
    this._pinSubir = pinSubir;
    this._pinDescer = pinDescer;
    this._motorEsquerda = motorEsquerda;
    this._motorDireita = motorDireita;
    this._pig = pig;
  }

  async begin(): Promise<void> {
    // Botões: pull-down → nível HIGH quando pressionados (ligados ao 3.3V)
    this._gpioBtnSubir  = this._pig.gpio(this._pinSubir);
    this._gpioBtnDescer = this._pig.gpio(this._pinDescer);

    await this._gpioBtnSubir.modeSet('input');
    await this._gpioBtnSubir.pullUpDown(1);  // PUD_DOWN = 1

    await this._gpioBtnDescer.modeSet('input');
    await this._gpioBtnDescer.pullUpDown(1);

    // Lê estado inicial dos botões
    this._btnSubirLevel  = await this._gpioBtnSubir.read();
    this._btnDescerLevel = await this._gpioBtnDescer.read();

    // Mantém estado atualizado via notify (sem polling)
    this._gpioBtnSubir.notify((level: number) => { this._btnSubirLevel = level; });
    this._gpioBtnDescer.notify((level: number) => { this._btnDescerLevel = level; });

    // Loop de leitura de botões + controle: 5ms (200 Hz)
    this._pollInterval = setInterval(() => this._loopStep(), 5);

    // Loop de sincronização: 20ms (50 Hz)
    this._sincInterval = setInterval(() => this._sincronizarMotores(), SINC.INTERVALO_MS);
  }

  subir(): void {
    this._salvarStepsInicio();
    this._mqttComando = 'SUBIR';
    this._modoPreset = false;
    this._motorEsquerda.setVelocidade(MOTOR.VEL_DEFAULT);
    this._motorDireita.setVelocidade(MOTOR.VEL_DEFAULT);
    this._motorEsquerda.subir();
    this._motorDireita.subir();
  }

  descer(): void {
    this._salvarStepsInicio();
    this._mqttComando = 'DESCER';
    this._modoPreset = false;
    this._motorEsquerda.setVelocidade(MOTOR.VEL_DEFAULT);
    this._motorDireita.setVelocidade(MOTOR.VEL_DEFAULT);
    this._motorEsquerda.descer();
    this._motorDireita.descer();
  }

  parar(): void {
    this._mqttComando = 'PARAR';
    this._modoPreset = false;
    this._motorEsquerda.parar();
    this._motorDireita.parar();
    this._salvarStepsJson().catch((e) => console.error('[Controles] Erro ao salvar steps:', e));
  }

  irPara(alvoEmCm: number): void {
    if (alvoEmCm < PRESET.ALTURA_MIN_CM || alvoEmCm > PRESET.ALTURA_MAX_CM) {
      console.warn(`[Controles] Preset fora de faixa: ${alvoEmCm} cm`);
      return;
    }
    this._alvoEmCm = alvoEmCm;
    this._modoPreset = true;
    this._salvarStepsInicio();
  }

  async calibrar(mqtt: MqttClient): Promise<void> {
    if (this._calibrando) return;
    this._calibrando = true;
    this._mqttComando = 'PARAR';
    this._modoPreset = false;

    try {
      console.log('[Calibracao] Iniciando — descendo...');
      mqtt.publishInfo({ tipo: 'status', status: 'calibrando', fase: 'descendo' });

      // Fase 1: descer até travar
      this._motorEsquerda.setVelocidade(MOTOR.VEL_DEFAULT);
      this._motorDireita.setVelocidade(MOTOR.VEL_DEFAULT);
      this._motorEsquerda.descer();
      this._motorDireita.descer();

      await _sleep(2000);
      await this._waitUntilStalled(MOTOR.ENCODER_TIMEOUT_MS);

      this._motorEsquerda.parar();
      this._motorDireita.parar();
      await _sleep(300);

      // Zera encoders na posição mínima
      this._motorEsquerda.setSteps(0);
      this._motorDireita.setSteps(0);

      // Fase 2: subir até travar
      console.log('[Calibracao] Subindo para medir...');
      mqtt.publishInfo({ tipo: 'status', status: 'calibrando', fase: 'subindo' });

      await _sleep(500);
      this._motorEsquerda.subir();
      this._motorDireita.subir();

      await _sleep(2000);
      await this._waitUntilStalled(MOTOR.ENCODER_TIMEOUT_MS);

      const pulsosE = this._motorEsquerda.getSteps();
      const pulsosD = this._motorDireita.getSteps();

      this._motorEsquerda.parar();
      this._motorDireita.parar();

      // Salva calibração
      this._motorEsquerda.setPulsosCalibrados(pulsosE);
      this._motorDireita.setPulsosCalibrados(pulsosD);

      await this._salvarStepsJson();

      console.log(`[Calibracao] Concluída — E:${pulsosE} D:${pulsosD}`);
      mqtt.publishCalibration({ pulsosE, pulsosD });
    } finally {
      this._calibrando = false;
    }
  }

  getStatusPayload(): StatusPayload {
    const direcaoE = this._motorEsquerda.getDirecao();
    const posicao  = this._motorEsquerda.getPosicaoEmCentimetros();

    let status: string;
    if (direcaoE === 'SUBIR')       status = 'subindo';
    else if (direcaoE === 'DESCER') status = 'descendo';
    else                            status = 'parado';

    return {
      status,
      posicao:  Math.round(posicao * 100) / 100,
      stepsE:   this._motorEsquerda.getSteps(),
      stepsD:   this._motorDireita.getSteps(),
      erroSync: this._erroSincAtual,
      velE:     this._motorEsquerda.getVelocidade(),
      velD:     this._motorDireita.getVelocidade(),
      btnS:     this._isBotaoSubir(),
      btnD:     this._isBotaoDescer(),
    };
  }

  teardown(): void {
    if (this._pollInterval) clearInterval(this._pollInterval);
    if (this._sincInterval) clearInterval(this._sincInterval);
    this._motorEsquerda.parar();
    this._motorDireita.parar();
    try { this._gpioBtnSubir?.endNotify(); } catch { /* noop */ }
    try { this._gpioBtnDescer?.endNotify(); } catch { /* noop */ }
  }

  // ---------------------------------------------------------------------------
  // Loop principal (5ms)
  // ---------------------------------------------------------------------------

  private _loopStep(): void {
    if (this._calibrando) return;

    const btnSubir  = this._isBotaoSubir();
    const btnDescer = this._isBotaoDescer();

    // Botão físico cancela recovery e preset
    if (this._recuperando && (btnSubir || btnDescer)) {
      this._recuperando = false;
      this._motorEsquerda.parar();
      this._motorDireita.parar();
      return;
    }

    // Modo recovery: retorna à posição inicial com tolerância ±5 steps
    if (this._recuperando) {
      const stepsE = this._motorEsquerda.getSteps();
      const diff = stepsE - this._stepsEInicio;
      if (Math.abs(diff) <= 5) {
        this._recuperando = false;
        this._motorEsquerda.parar();
        this._motorDireita.parar();
        this._salvarStepsJson().catch(() => {});
      } else if (diff > 5) {
        this._motorEsquerda.descer();
        this._motorDireita.descer();
      } else {
        this._motorEsquerda.subir();
        this._motorDireita.subir();
      }
      return;
    }

    // Botão físico cancela preset e executa manualmente
    if (btnSubir || btnDescer) {
      if (this._modoPreset) {
        this._modoPreset = false;
        this._motorEsquerda.parar();
        this._motorDireita.parar();
        return;
      }

      if (btnSubir && !btnDescer) {
        if (this._mqttComando !== 'SUBIR') {
          this._salvarStepsInicio();
          this._mqttComando = 'PARAR';
        }
        this._motorEsquerda.subir();
        this._motorDireita.subir();
      } else if (btnDescer && !btnSubir) {
        if (this._mqttComando !== 'DESCER') {
          this._salvarStepsInicio();
          this._mqttComando = 'PARAR';
        }
        this._motorEsquerda.descer();
        this._motorDireita.descer();
      } else {
        // Ambos pressionados → para
        this._motorEsquerda.parar();
        this._motorDireita.parar();
      }
      return;
    }

    // Modo preset: busca altura alvo
    if (this._modoPreset) {
      const posAtual = this._motorEsquerda.getPosicaoEmCentimetros();
      const diff = this._alvoEmCm - posAtual;

      if (Math.abs(diff) <= PRESET.TOLERANCIA_CM) {
        this._modoPreset = false;
        this._motorEsquerda.parar();
        this._motorDireita.parar();
        this._salvarStepsJson().catch(() => {});
      } else if (diff > 0) {
        this._motorEsquerda.subir();
        this._motorDireita.subir();
      } else {
        this._motorEsquerda.descer();
        this._motorDireita.descer();
      }
      return;
    }

    // Executa comando MQTT
    if (this._mqttComando === 'SUBIR') {
      this._motorEsquerda.subir();
      this._motorDireita.subir();
    } else if (this._mqttComando === 'DESCER') {
      this._motorEsquerda.descer();
      this._motorDireita.descer();
    } else {
      this._motorEsquerda.parar();
      this._motorDireita.parar();
    }
  }

  // ---------------------------------------------------------------------------
  // Sincronização proporcional (20ms)
  // ---------------------------------------------------------------------------

  private _sincronizarMotores(): void {
    if (this._calibrando || this._recuperando) return;

    const direcaoE = this._motorEsquerda.getDirecao();
    const direcaoD = this._motorDireita.getDirecao();
    if (direcaoE === 'PARAR' && direcaoD === 'PARAR') return;

    const stepsE = this._motorEsquerda.getSteps();
    const stepsD = this._motorDireita.getSteps();
    const pulsosE = this._motorEsquerda.getPulsosCalibrados();
    const pulsosD = this._motorDireita.getPulsosCalibrados();

    if (pulsosE === 0 || pulsosD === 0) {
      console.error('[Sinc] pulsosCalibrados inválido — sincronização abortada');
      return;
    }

    const deltaE = (stepsE - this._stepsEInicio) / pulsosE;
    const deltaD = (stepsD - this._stepsDInicio) / pulsosD;
    const erro   = deltaE - deltaD;

    const pulsosRef = (pulsosE + pulsosD) / 2;
    this._erroSincAtual = Math.round(erro * pulsosRef);

    const encoderInativo =
      stepsE === this._stepsEInicio || stepsD === this._stepsDInicio;

    if (Math.abs(this._erroSincAtual) > SINC.EMERGENCIA_PULSOS) {
      this._motorEsquerda.parar();
      this._motorDireita.parar();
      this._recuperando = true;
      const msg = encoderInativo ? 'encoder_inativo' : 'sinc_emergencia';
      console.error(`[Sinc] Emergência: ${msg} — erroSync=${this._erroSincAtual}`);
      return;
    }

    if (erro > SINC.THRESHOLD) {
      const velE = Math.max(SINC.VEL_MIN, Math.min(255, Math.round(255 - SINC.GANHO * erro)));
      this._motorEsquerda.setVelocidade(velE);
      this._motorDireita.setVelocidade(MOTOR.VEL_DEFAULT);
    } else if (erro < -SINC.THRESHOLD) {
      const velD = Math.max(SINC.VEL_MIN, Math.min(255, Math.round(255 + SINC.GANHO * erro)));
      this._motorEsquerda.setVelocidade(MOTOR.VEL_DEFAULT);
      this._motorDireita.setVelocidade(velD);
    } else {
      this._motorEsquerda.setVelocidade(MOTOR.VEL_DEFAULT);
      this._motorDireita.setVelocidade(MOTOR.VEL_DEFAULT);
    }
  }

  // ---------------------------------------------------------------------------
  // Helpers
  // ---------------------------------------------------------------------------

  private _salvarStepsInicio(): void {
    this._stepsEInicio = this._motorEsquerda.getSteps();
    this._stepsDInicio = this._motorDireita.getSteps();
  }

  private _isBotaoSubir(): boolean {
    return this._btnSubirLevel === 1;
  }

  private _isBotaoDescer(): boolean {
    return this._btnDescerLevel === 1;
  }

  private _waitUntilStalled(timeoutMs: number, maxWaitMs = 90_000): Promise<void> {
    return new Promise((resolve, reject) => {
      const start = Date.now();
      const check = setInterval(() => {
        if (this._motorEsquerda.isTravado(timeoutMs) && this._motorDireita.isTravado(timeoutMs)) {
          clearInterval(check);
          resolve();
        } else if (Date.now() - start > maxWaitMs) {
          clearInterval(check);
          reject(new Error('Timeout: motor não travou dentro do tempo esperado'));
        }
      }, 50);
    });
  }

  private async _salvarStepsJson(): Promise<void> {
    const data = {
      esquerda: {
        pulsosCalibrados: this._motorEsquerda.getPulsosCalibrados(),
        steps: this._motorEsquerda.getSteps(),
      },
      direita: {
        pulsosCalibrados: this._motorDireita.getPulsosCalibrados(),
        steps: this._motorDireita.getSteps(),
      },
    };
    await fs.writeFile(CALIBRACAO_PATH, JSON.stringify(data, null, 2), 'utf-8');
  }
}

function _sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}
